{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4efa437-a86c-4dc4-bf39-c5a3ed1ab2aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "library(ArchR)\n",
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(GenomicRanges)\n",
    "library(future)\n",
    "library(\"EnsDb.Hsapiens.v86\")\n",
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "library(dplyr)\n",
    "library(igraph)\n",
    "library(tidyverse)\n",
    "library(patchwork)\n",
    "library(viridis)\n",
    "library(scCustomize)\n",
    "library(qs)\n",
    "library(ggpubr)\n",
    "library(clustree)\n",
    "library(\"ggraph\")\n",
    "library(\"Nebulosa\")\n",
    "library(ggalluvial)\n",
    "library(\"MAST\")\n",
    "library(\"pheatmap\")\n",
    "library(\"chromVAR\")\n",
    "library(\"TFBSTools\")\n",
    "#library(\"JASPAR2022\")\n",
    "library(\"motifmatchr\")\n",
    "library(\"presto\")\n",
    "library(\"JASPAR2022\")\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(ChIPseeker)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(Seurat)\n",
    "library(sctransform)\n",
    "library(dplyr)\n",
    "#library(scDblFinder)\n",
    "library(biomaRt) \n",
    "#library(scater)\n",
    "library(SingleCellExperiment)\n",
    "#library(SeuratDisk)\n",
    "library(Signac)\n",
    "library(GenomicRanges)\n",
    "library(rtracklayer)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(scales)\n",
    "library(future) \n",
    "#plan(\"multicore\", workers = 10)\n",
    "#options(future.globals.maxSize = 1000000 * 1024^2)\n",
    "library(harmony)\n",
    "library(biovizBase)\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce112be8-10eb-4983-bdaa-f1e324eeec75",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (!dir.exists(\"Images\")) {dir.create(\"Images\")}\n",
    "if (!dir.exists(\"Tables\"))  {dir.create(\"Tables\")}\n",
    "if (!dir.exists(\"RObjects\")) {dir.create(\"RObjects\")}\n",
    "\n",
    "IMG_OUT = \"Images\"\n",
    "DF_OUT  = \"Tables\"\n",
    "RDS_OUT = \"RObjects\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc172b55-b439-43a0-88bd-7cee0cbf7e79",
   "metadata": {},
   "outputs": [],
   "source": [
    "#basic steps to read in data and performing QC\n",
    "    ##Creating a common peak set\n",
    "    ### read in peak sets\n",
    "    ### convert to genomic ranges\n",
    "    ### Create a unified set of peaks to quantify in each dataset\n",
    "    ### Filter out bad peaks based on length\n",
    "\n",
    "    ##Create Fragment objects\n",
    "    ### load metadata\n",
    "    ### perform an initial filtering of low count cells\n",
    "    ### create fragment objects\n",
    "\n",
    "    ##Quantify peaks in each dataset\n",
    "    \n",
    "    ##Create the objects\n",
    "\n",
    "    ##Annotate using EnsDb.Hsapiens.v86\n",
    "\n",
    "    ##Perform basic QC based on nCount_ATAC, TSS.Enrichment and Nucleosomal.Signal\n",
    "    \n",
    "    ##Integrate datasets to be analyzed together\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00c5a0a0-8b4f-446d-a427-6bc60462b156",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add information to identify dataset of origin\n",
    "tscm.l1$Origin <- 'tscm.l1'\n",
    "tscm.l2$Origin <- 'tscm.l2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7c19826-9c21-4bd3-b32f-86ce5e3266de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute LSI\n",
    "tscm.l1 <- FindTopFeatures(tscm.l1, min.cutoff = 10)\n",
    "tscm.l1 <- RunTFIDF(tscm.l1)\n",
    "tscm.l1 <- RunSVD(tscm.l1)\n",
    "\n",
    "# compute LSI\n",
    "tscm.l2 <- FindTopFeatures(tscm.l2, min.cutoff = 10)\n",
    "tscm.l2 <- RunTFIDF(tscm.l2)\n",
    "tscm.l2 <- RunSVD(tscm.l2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90afe3a1-56e8-4e8b-aef2-e8490882e2fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find integration anchors\n",
    "integration.anchors <- FindIntegrationAnchors(\n",
    "  object.list = list(tscm.l1, tscm.l2),\n",
    "  anchor.features = rownames(tscm.l2),\n",
    "  reduction = \"rlsi\",\n",
    "  dims = 2:30\n",
    ")\n",
    "\n",
    "# integrate LSI embeddings\n",
    "integrated <- IntegrateEmbeddings(\n",
    "  anchorset = integration.anchors,\n",
    "  reductions = combined[[\"lsi\"]],\n",
    "  new.reduction.name = \"integrated_lsi\",\n",
    "  dims.to.integrate = 1:30\n",
    ")\n",
    "\n",
    "# create a new UMAP using the integrated embeddings\n",
    "integrated <- RunUMAP(integrated, reduction = \"integrated_lsi\", dims = 2:30, verbose = FALSE)\n",
    "p2 <- DimPlot(integrated, group.by = \"Origin\")\n",
    "p2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fda5a9bd-08f9-4ad6-b312-8fd5ee424e29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gene annotations for hg38\n",
    "suppressWarnings({\n",
    "annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)\n",
    "seqlevels(annotation) <- paste0('chr', seqlevels(annotation))\n",
    "seqlevelsStyle(annotation) <- 'UCSC'\n",
    "\n",
    "Annotation(object = integrated) <- annotation\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca8bc4b0-5391-452a-9b51-20f5dd5c2cb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# call peaks using MACS2\n",
    "peaks <- CallPeaks(seu.integrated)\n",
    "\n",
    "# remove peaks on nonstandard chromosomes and in genomic blacklist regions\n",
    "peaks <- keepStandardChromosomes(peaks, pruning.mode = \"coarse\")\n",
    "peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)\n",
    "\n",
    "# quantify counts in each peak\n",
    "macs2_counts <- FeatureMatrix(\n",
    "  fragments = Fragments(seu.integrated),\n",
    "  features = peaks,\n",
    "  cells = colnames(seu.integrated)\n",
    ")\n",
    "\n",
    "# create a new assay using the MACS2 peak set and add it to the Seurat object\n",
    "seu.integrated[[\"peaks\"]] <- CreateChromatinAssay(\n",
    "  counts = macs2_counts,\n",
    "  annotation = annotation,\n",
    "  fragments = Fragments(seu.integrated)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c02300c6-3f4c-456d-98fb-8c76f2aaff32",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use MACS2-called \"peaks\" assay to perform downstream analysis\n",
    "DefaultAssay(seu.integrated) <- \"peaks\"\n",
    "seu.integrated <- RunTFIDF(seu.integrated)\n",
    "seu.integrated <- FindTopFeatures(seu.integrated, min.cutoff = 10)\n",
    "seu.integrated <- RunSVD(seu.integrated)\n",
    "seu.integrated <- RunUMAP(seu.integrated, dims = 2:30, reduction = 'integrated_lsi', metric = \"euclidean\", verbose = FALSE)\n",
    "\n",
    "seu.integrated <- FindNeighbors(object = seu.integrated, reduction = 'integrated_lsi', dims = 2:30, graph.name = c(\"peaks_nn\", \"peaks_snn\"))\n",
    "\n",
    "seu.integrated <- FindClusters(object = seu.integrated, verbose = FALSE, algorithm = 3, resolution = 0.4, graph.name = \"peaks_snn\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f1715ae-0291-47dd-8430-98c6bd03178e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate umaps (Figure 3E)\n",
    "Idents(seu.integrated) <- \"peaks_snn_res.0.4\"\n",
    "pdf(sprintf(\"%s/umap.integrated.pdf\", IMG_OUT), width=6, height=5)\n",
    "DimPlot(seu.integrated, label = TRUE, repel = TRUE)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29e38647-6c26-4203-9e71-6001d47ef002",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a gene activity matrix\n",
    "\n",
    "gene.activities <- GeneActivity(seu.integrated)\n",
    "\n",
    "# add the gene activity matrix to the Seurat object as a new assay and normalize it\n",
    "seu.integrated[['RNA']] <- CreateAssayObject(counts = gene.activities)\n",
    "seu.integrated <- NormalizeData(\n",
    "  object = seu.integrated,\n",
    "  assay = 'RNA',\n",
    "  normalization.method = 'LogNormalize',\n",
    "  scale.factor = median(seu.integrated$nCount_RNA)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0421a0c7-372f-43c7-9444-57f16dd04ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prediction of scRNA-based clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5d202ee-aacb-4a56-8b54-877c32992155",
   "metadata": {},
   "outputs": [],
   "source": [
    "transfer.anchors <- FindTransferAnchors(reference = seu.rna, query = seu.integrated, features = VariableFeatures(object = seu.rna), \n",
    "    reference.assay = \"RNA\", query.assay = \"RNA\", reduction = \"cca\")\n",
    "\n",
    "celltype.predictions <- TransferData(anchorset = transfer.anchors, refdata = seu.rna$SCT_snn_res.0.4, \n",
    "    weight.reduction = seu.integrated[[\"integrated_lsi\"]], dims = 2:30)\n",
    "\n",
    "seu.integrated <- AddMetaData(seu.integrated, metadata = celltype.predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de6f17f1-abbf-4484-8439-15da0b515529",
   "metadata": {},
   "outputs": [],
   "source": [
    "#paint clustering information of scRNA data on predicted scATAC data (Figure 3G)\n",
    "Idents(seu.integrated) <- \"predicted.id\"\n",
    "seu.integrated@meta.data %>% filter(predicted.id %in% c(\"11\")) %>% rownames() -> cl11.bcs\n",
    "seu.integrated@meta.data %>% filter(predicted.id %in% c(\"12\")) %>% rownames() -> cl12.bcs\n",
    "\n",
    "#pdf(sprintf(\"%s/umap.predicted.id.cl11.12.highlight.pdf\", IMG_OUT), width=7, height=5)\n",
    "DimPlot(seu.integrated, cells.highlight = list(cl11.bcs, cl12.bcs), cols.highlight = c(\"#0f9ed5\", \"#78206e\"), pt.size = 2, sizes.highlight = 2)\n",
    "#dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "923ee340-4cf2-4dbf-bff4-eedd201a100d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate cicero-based links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87e1c486-5beb-4315-ac52-695c5d6275f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "seu.integrated[[\"UMAP\"]] <- seu.integrated[[\"umap\"]]\n",
    "int.cds <- as.cell_data_set(x = seu.integrated)\n",
    "umap_coords <- reducedDims(int.cds)$UMAP\n",
    "int.cicero <- make_cicero_cds(int.cds, reduced_coordinates = umap_coords)\n",
    "\n",
    "# get the chromosome sizes from the Seurat object\n",
    "genome <- seqlengths(seu.integrated@assays$peaks@annotation)\n",
    "\n",
    "# convert chromosome sizes to a dataframe\n",
    "genome.df <- data.frame(\"chr\" = names(genome), \"length\" = genome)\n",
    "\n",
    "# run cicero\n",
    "conns <- run_cicero(int.cicero, genomic_coords = genome.df, sample_num = 100)\n",
    "\n",
    "ccans <- generate_ccans(conns)\n",
    "\n",
    "links <- ConnectionsToLinks(conns = conns, ccans = ccans)\n",
    "Links(seu.integrated) <- links\n",
    "\n",
    "Links(seu.integrated) %>% as.data.frame() %>% write.csv(\"cicero.links.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d38799c4-a99e-49cb-b835-a006ff29d329",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate coverage plots (Figures 3H, 6G-H, S3D and S6A-B)\n",
    "##coverage across clusters for genomic loci of interest\n",
    "##violin plots (right) for gene activity for gene of interest\n",
    "##peak plot corresponding to scATAC data\n",
    "##Gene annotation for loci visualized\n",
    "##ChIP-seq data as peak plot or BigWig track from publicly available dataset\n",
    "##cicero-generated links\n",
    "##H3K27ac from ENCODE\n",
    "##H3K4me1 from ENCODE\n",
    "##BigWig tracks from bulk-ATAC-Seq (10 donors) (only Figures 6G-6H and S6A-B)\n",
    "##Peaks containing motif for TFs on interest (only Figure 6H)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67db05ad-58fb-49cd-81f5-d871db7fd781",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gzmb (Figure 3H)\n",
    "roi = \"chr14-24600000-24645000\"\n",
    "cov_plot <- CoveragePlot(\n",
    "  object = seu.integrated,\n",
    "  region = roi,\n",
    "  annotation = FALSE,\n",
    "  peaks = FALSE,\n",
    "  links = FALSE\n",
    ")\n",
    "#cov_plot\n",
    "\n",
    "gene_plot <- AnnotationPlot(\n",
    "  object = seu.integrated,\n",
    "  region = roi\n",
    ")\n",
    "#gene_plot\n",
    "\n",
    "peak_plot1 <- PeakPlot(\n",
    "  object = seu.integrated,\n",
    "  region = roi\n",
    ")\n",
    "#peak_plot1\n",
    "\n",
    "peak_plot2 <- PeakPlot(seu.integrated, \n",
    "         region = roi, \n",
    "         peaks = GRanges(import(\"../eomes_chip_hg38.bed\")), \n",
    "         color = \"#DC268C\")\n",
    "#peak_plot2\n",
    "\n",
    "link_plot <- LinkPlot(\n",
    "  object = seu.integrated,\n",
    "  region = roi\n",
    ")\n",
    "#link_plot\n",
    "\n",
    "expr_plot <- ExpressionPlot(\n",
    "  object = seu.integrated,\n",
    "  features = \"GZMB\",\n",
    "  assay = \"RNA\"\n",
    ")\n",
    "#expr_plot\n",
    "\n",
    "\n",
    "chip2 <- BigwigTrack(region = roi, \n",
    "                     \"../tbet_chip3_hg38.bigWig\", \n",
    "                     ymax = \"q95\",\n",
    "                     y_label = \"T-bet\") + scale_fill_manual(values=\"#921B63\") + theme(legend.position=\"none\")\n",
    "chip3 <- BigwigTrack(region = roi, \n",
    "                     \"../fosl2_r1.bigWig\", \n",
    "                     ymax = \"q95\",\n",
    "                     y_label = \"FOSL2\") + scale_fill_manual(values=\"#FA3939\") + theme(legend.position=\"none\")\n",
    "\n",
    "\n",
    "chip5 <- BigwigTrack(roi, \n",
    "                     \"../ENCODE/H3K27ac_CD4abT/1/ENCFF884NBE.bigWig\",\n",
    "                     y_label = \"H3K27ac\",\n",
    "                     smooth = 100) + theme(legend.position=\"none\") + scale_fill_manual(values=\"black\")\n",
    "\n",
    "chip6 <- BigwigTrack(roi, \n",
    "                     \"../ENCODE/H3K4me1_CD45RO_CD4/ENCFF334TZP.bigWig\",\n",
    "                     y_label = \"H3K4me1\",\n",
    "                     smooth = 100) + theme(legend.position=\"none\") + scale_fill_manual(values=\"darkgrey\")\n",
    "\n",
    "CombineTracks(\n",
    "  plotlist = list(cov_plot, peak_plot1, gene_plot, peak_plot2, chip2, chip3, link_plot, chip5, chip6),\n",
    "  expression.plot = expr_plot,\n",
    "  heights = c(10, 1, 2, 1, 2, 2, 3, 2, 2),\n",
    "  widths = c(10, 1)\n",
    ") -> gzmb.cov\n",
    "gzmb.cov\n",
    "\n",
    "pdf(sprintf(\"%s/gzmb.cov.pdf\", IMG_OUT), width=8, height=8)\n",
    "gzmb.cov\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0b30930-c225-4eaa-a842-e5d889077e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#slamf7 (Figures 6G and S6A)\n",
    "cov_plot <- CoveragePlot(\n",
    "  object = seu.filt,\n",
    "  region = roi4,\n",
    "  annotation = FALSE,\n",
    "  peaks = FALSE,\n",
    "  links = FALSE\n",
    ") & scale_fill_manual(values = c(\"#78206e\", \"#0f9ed5\", \"#FD95F8\", \"#80BB17\", \"#6C244C\"))\n",
    "#cov_plot\n",
    "\n",
    "gene_plot <- AnnotationPlot(\n",
    "  object = seu.filt,\n",
    "  region = roi4\n",
    ")\n",
    "#gene_plot\n",
    "\n",
    "peak_plot1 <- PeakPlot(\n",
    "  object = seu.filt,\n",
    "  region = roi4\n",
    ")\n",
    "#peak_plot1\n",
    "\n",
    "peak_plot2 <- PeakPlot(seu.filt, \n",
    "         region = roi4, \n",
    "         peaks = GRanges(import(\"../eomes_chip_hg38.bed\")), \n",
    "         color = \"#DC268C\")\n",
    "#peak_plot2\n",
    "\n",
    "link_plot <- LinkPlot(\n",
    "  object = seu.filt,\n",
    "  region = roi4,\n",
    "  min.cutoff = 0.05\n",
    ")\n",
    "#link_plot\n",
    "\n",
    "expr_plot <- ExpressionPlot(\n",
    "  object = seu.filt,\n",
    "  features = \"SLAMF7\",\n",
    "  assay = \"RNA\"\n",
    ") & scale_fill_manual(values = c(\"#78206e\", \"#0f9ed5\", \"#FD95F8\", \"#80BB17\", \"#6C244C\"))\n",
    "#expr_plot\n",
    "\n",
    "peak_plot3 <- PeakPlot(seu.filt, \n",
    "         region = roi4, \n",
    "         peaks = bulk.gr, \n",
    "         color = \"black\")\n",
    "\n",
    "gh <- PeakPlot(seu.filt, \n",
    "         region = roi4, \n",
    "         peaks = genehancer,\n",
    "         group.by = \"type\", \n",
    "         color = c(\"#DE6012\", \"#5087A0\"))\n",
    "\n",
    "chip2 <- BigwigTrack(region = roi4, \n",
    "                     list(\"Tbet\" = \"../tbet_chip3_hg38.bigWig\"), \n",
    "                     ymax = \"q95\") + scale_fill_manual(values=\"#921B63\") + theme(legend.position=\"none\")\n",
    "chip3 <- BigwigTrack(region = roi4, \n",
    "                     list(\"FOSL2\" = \"../fosl2_r1.bigWig\"), \n",
    "                     ymax = \"q95\") + scale_fill_manual(values=\"#FA3939\") + theme(legend.position=\"none\")\n",
    "\n",
    "bulk <- BigwigTrack(region = roi4, \n",
    "                     list(\n",
    "               \"naive\" = \"../bulk_bw/N_out.bw\",\n",
    "               \"temra_e\" = \"../bulk_bw/E_out.bw\"), \n",
    "                     ymax = 5000) + scale_fill_manual(values=c(\"#7F7F7F\", \"#F8766D\")) + theme(legend.position=\"none\")\n",
    "\n",
    "chip5 <- BigwigTrack(roi4, \n",
    "                     \"../ENCODE/H3K27ac_CD4abT/1/ENCFF884NBE.bigWig\",\n",
    "                     y_label = \"H3K27ac\",\n",
    "                     smooth = 100) + theme(legend.position=\"none\") + scale_fill_manual(values=\"black\")\n",
    "\n",
    "chip6 <- BigwigTrack(roi4, \n",
    "                     \"../ENCODE/H3K4me1_CD45RO_CD4/ENCFF334TZP.bigWig\",\n",
    "                     y_label = \"H3K4me1\",\n",
    "                     smooth = 100) + theme(legend.position=\"none\") + scale_fill_manual(values=\"darkgrey\")\n",
    "#chip5\n",
    "\n",
    "CombineTracks(\n",
    "  plotlist = list(cov_plot, peak_plot1, link_plot, bulk,\n",
    "                  peak_plot3, gene_plot, peak_plot2, chip2, chip3, chip5, chip6),\n",
    "  expression.plot = expr_plot,\n",
    "  heights = c(15, 1, 3, 6, 1, 2, 1, 2, 2, 2, 2),\n",
    "  widths = c(10, 1)\n",
    ") -> slamf7.cov\n",
    "slamf7.cov\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd3db1de-46d3-490b-9a80-6ab5496b96c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract motif containing peaks for TF of interest\n",
    "pfm2 <- readJASPARMatrix(\"../CellRangerReferences/refdata-cellranger-arc-GRCh38-2020-A-2.0.0/regions/motifs.pfm\", matrixClass=\"PFM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed25d35-aa95-4c2e-9e5e-2f18aefe61c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pfm.stat4 <- pfm2[grepl(\"STAT4\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.nfat5 <- pfm2[grepl(\"NFAT5\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.rel <- pfm2[grepl(\"REL_\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.arntl <- pfm2[grepl(\"ARNTL\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.klf12 <- pfm2[grepl(\"KLF12\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.tcf12 <- pfm2[grepl(\"TCF12_\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.foxn3 <- pfm2[grepl(\"FOXN3\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.nfkb1 <- pfm2[grepl(\"NFKB1\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.cux1 <- pfm2[grepl(\"CUX1\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.runx2 <- pfm2[grepl(\"RUNX2\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.nr3c1 <- pfm2[grepl(\"NR3C1\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.smad3 <- pfm2[grepl(\"SMAD3\", names(pfm2), ignore.case = TRUE)]\n",
    "pfm.nfatc3 <- pfm2[grepl(\"NFATC3\", names(pfm2), ignore.case = TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db719c3e-067d-49de-990d-807d6a78e11b",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_pos_stat4 <- matchMotifs(pfm.stat4, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_nfat5 <- matchMotifs(pfm.nfat5, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_rel <- matchMotifs(pfm.rel, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_arntl <- matchMotifs(pfm.arntl, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_klf12 <- matchMotifs(pfm.klf12, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_tcf12 <- matchMotifs(pfm.tcf12, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_foxn3 <- matchMotifs(pfm.foxn3, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_nfkb1 <- matchMotifs(pfm.nfkb1, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_cux1 <- matchMotifs(pfm.cux1, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_runx2 <- matchMotifs(pfm.runx2, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_nr3c1 <- matchMotifs(pfm.nr3c1, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_smad3 <- matchMotifs(pfm.smad3, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)\n",
    "\n",
    "motif_pos_nfatc3 <- matchMotifs(pfm.nfatc3, StringToGRanges(rownames(seu.filt[[\"peaks\"]])), \n",
    "                                   genome = BSgenome.Hsapiens.UCSC.hg38,  out=\"positions\", p.cutoff=0.0005)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5cc46d2-37a7-47d5-8430-3224eeb9ba56",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_pos_stat4$Stat4_MA0518.1 %>% filter(score>10) -> motif_pos_stat4_filt1\n",
    "motif_pos_nfat5$NFAT5_MA0606.1 %>% filter(score>10) -> motif_pos_nfat5_filt1\n",
    "motif_pos_rel$REL_MA0101.1 %>% filter(score>10) -> motif_pos_rel_filt1\n",
    "motif_pos_arntl$Arntl_MA0603.1 %>% filter(score>10) -> motif_pos_arntl_filt1\n",
    "motif_pos_klf12$Klf12_MA0742.1 %>% filter(score>10) -> motif_pos_klf12_filt1\n",
    "motif_pos_tcf12$Tcf12_MA0521.1 %>% filter(score>10) -> motif_pos_tcf12_filt1\n",
    "motif_pos_foxn3$FOXN3_MA1489.1 %>% filter(score>10) -> motif_pos_foxn3_filt1\n",
    "motif_pos_nfkb1$NFKB1_MA0105.4 %>% filter(score>10) -> motif_pos_nfkb1_filt1\n",
    "motif_pos_cux1$CUX1_MA0754.1 %>% filter(score>10) -> motif_pos_cux1_filt1\n",
    "motif_pos_runx2$RUNX2_MA0511.2 %>% filter(score>10) -> motif_pos_runx2_filt1\n",
    "motif_pos_nr3c1$NR3C1_MA0113.3 %>% filter(score>10) -> motif_pos_nr3c1_filt1\n",
    "motif_pos_smad3$SMAD3_MA0795.1 %>% filter(score>10) -> motif_pos_smad3_filt1\n",
    "motif_pos_nfatc3$NFATC3_MA0625.1 %>% filter(score>10) -> motif_pos_nfatc3_filt1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c072c942-c914-444c-aeb8-b73e04b5619d",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_pos_stat4_filt1$tf <- \"Stat4\"\n",
    "motif_pos_nfat5_filt1$tf <- \"NFAT5\"\n",
    "motif_pos_rel_filt1$tf <- \"REL\"\n",
    "motif_pos_arntl_filt1$tf <- \"Arntl\"\n",
    "motif_pos_klf12_filt1$tf <- \"Klf12\"\n",
    "motif_pos_tcf12_filt1$tf <- \"Tcf12\"\n",
    "motif_pos_foxn3_filt1$tf <- \"FOXN3\"\n",
    "motif_pos_nfkb1_filt1$tf <- \"NFKB1\"\n",
    "motif_pos_cux1_filt1$tf <- \"CUX1\"\n",
    "motif_pos_runx2_filt1$tf <- \"RUNX2\"\n",
    "motif_pos_nr3c1_filt1$tf <- \"NR3C1\"\n",
    "motif_pos_smad3_filt1$tf <- \"SMAD3\"\n",
    "motif_pos_nfatc3_filt1$tf <- \"NFATC3\"\n",
    "\n",
    "c(motif_pos_stat4_filt1, motif_pos_nfat5_filt1, motif_pos_rel_filt1, motif_pos_arntl_filt1,\n",
    "  motif_pos_klf12_filt1, motif_pos_tcf12_filt1, motif_pos_foxn3_filt1, motif_pos_nfkb1_filt1,\n",
    "  motif_pos_cux1_filt1, motif_pos_runx2_filt1, motif_pos_nr3c1_filt1, motif_pos_smad3_filt1,\n",
    "  motif_pos_nfatc3_filt1) -> ranges.ofint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72771abb-4159-45bc-b545-5959bccf4ba2",
   "metadata": {},
   "outputs": [],
   "source": [
    "require(purrr)\n",
    "require(stringr)\n",
    "require(dplyr)\n",
    "require(ggplot2)\n",
    "\n",
    "# this randomly generates a color\n",
    "rand_color <- function(){\n",
    "  1:3 %>% \n",
    "    map_chr(~ str_pad(as.hexmode(floor(runif(1,0,256))),2,pad=\"0\")) %>%\n",
    "    paste0(collapse=\"\") %>% \n",
    "    paste0(\"#\",.)\n",
    "}\n",
    "\n",
    "# this is a function that, given a number, returns that number of colors\n",
    "my_colors <- function(n) 1:n %>% map_chr(~ rand_color())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54cfa35a-d9cc-4f37-9767-cb534bfd7366",
   "metadata": {},
   "outputs": [],
   "source": [
    "limits <- c(\"Stat4\",\"NFAT5\",\"REL\",\"Arntl\",\"Klf12\",\"Tcf12\",\"FOXN3\",\"NFKB1\",\"CUX1\",\"RUNX2\",\"NR3C1\",\"SMAD3\",\"NFATC3\")\n",
    "tfcols <- c('#f03111','#b3d1f8','#4d27a9','#0d52a1','#c21976','#464670','#b22a1e','#764907','#7dd615','#ea26d6','#4c9dfa','#9e538e','#0814f7')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0796d1fc-3eb3-450e-be18-20c645a0075b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fosl2 (Figures 6H and S6B)\n",
    "cov_plot <- CoveragePlot(\n",
    "  object = seu.filt,\n",
    "  region = roi8,\n",
    "  annotation = FALSE,\n",
    "  peaks = FALSE,\n",
    "  links = FALSE\n",
    ") & scale_fill_manual(values = c(\"#78206e\", \"#0f9ed5\", \"#FD95F8\", \"#80BB17\", \"#6C244C\"))\n",
    "#cov_plot\n",
    "\n",
    "gene_plot <- AnnotationPlot(\n",
    "  object = seu.filt,\n",
    "  region = roi8\n",
    ")\n",
    "#gene_plot\n",
    "\n",
    "peak_plot1 <- PeakPlot(\n",
    "  object = seu.filt,\n",
    "  region = roi8\n",
    ")\n",
    "#peak_plot1\n",
    "\n",
    "peak_plot2 <- PeakPlot(seu.filt, \n",
    "         region = roi8, \n",
    "         peaks = ranges.ofint, \n",
    "         group.by = \"tf\") & scale_color_manual(limits = limits, values = tfcols) & theme(legend.position=\"none\")\n",
    "#peak_plot2\n",
    "\n",
    "link_plot <- LinkPlot(\n",
    "  object = seu.filt,\n",
    "  region = roi8,\n",
    "  min.cutoff = 0.05\n",
    ")\n",
    "#link_plot\n",
    "\n",
    "expr_plot <- ExpressionPlot(\n",
    "  object = seu.filt,\n",
    "  features = \"FOSL2\",\n",
    "  assay = \"RNA\"\n",
    ") & scale_fill_manual(values = c(\"#78206e\", \"#0f9ed5\", \"#FD95F8\", \"#80BB17\", \"#6C244C\"))\n",
    "#expr_plot\n",
    "\n",
    "peak_plot3 <- PeakPlot(seu.filt, \n",
    "         region = roi8, \n",
    "         peaks = bulk.gr, \n",
    "         color = \"black\")\n",
    "\n",
    "bulk <- BigwigTrack(region = roi8, \n",
    "                     list(\n",
    "               \"naive\" = \"../bulk_bw/N_out.bw\",\n",
    "               \"temra_e\" = \"../bulk_bw/E_out.bw\"), \n",
    "                     ymax = 5000) + scale_fill_manual(values=c(\"#7F7F7F\", \"#F8766D\")) + theme(legend.position=\"none\")\n",
    "\n",
    "chip5 <- BigwigTrack(roi8, \n",
    "                     \"../ENCODE/H3K27ac_CD4abT/1/ENCFF884NBE.bigWig\",\n",
    "                     y_label = \"H3K27ac\",\n",
    "                     smooth = 100) + theme(legend.position=\"none\") + scale_fill_manual(values=\"black\")\n",
    "\n",
    "chip6 <- BigwigTrack(roi8, \n",
    "                     \"../ENCODE/H3K4me1_CD45RO_CD4/ENCFF334TZP.bigWig\",\n",
    "                     y_label = \"H3K4me1\",\n",
    "                     smooth = 100) + theme(legend.position=\"none\") + scale_fill_manual(values=\"darkgrey\")\n",
    "#chip5\n",
    "\n",
    "CombineTracks(\n",
    "  plotlist = list(cov_plot, peak_plot1, link_plot, bulk,\n",
    "                  peak_plot3, gene_plot, peak_plot2, chip5, chip6),\n",
    "  expression.plot = expr_plot,\n",
    "  heights = c(15, 1, 3, 6, 1, 2, 2, 2, 2),\n",
    "  widths = c(10, 1)\n",
    ") -> fosl2.cov\n",
    "fosl2.cov\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33980de4-8d90-4760-bdee-b635a5e744f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6390479-26a0-467a-9197-13a433f12386",
   "metadata": {},
   "outputs": [],
   "source": [
    "#module score analysis (Figures 3F, 6F and S3C)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b7e4c1-bfd5-4b16-af17-b90355f105dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "seu.integrated -> abc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "973ae8ae-98f3-4e51-bff5-332eaf1acca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "bulk.lists <- parse_marker_gmt(\"GMT/bulk.lists.narrowed.gmt\", trim_first=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19963e6d-f896-45b8-ada3-3ca84caf37ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(abc) <- \"peaks\"\n",
    "RegionStats(\n",
    "  object = abc,\n",
    "  genome = BSgenome.Hsapiens.UCSC.hg38,\n",
    "  assay = \"peaks\"\n",
    ") -> def\n",
    "\n",
    "suppressWarnings({\n",
    "LinkPeaks(def,\n",
    "         peak.assay = \"peaks\",\n",
    "         expression.assay = \"RNA\"\n",
    "         ) -> def\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c86b792a-5204-48d3-9b47-fc49d73938d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "GetLinkedPeaks(def, features = bulk.lists[[\"EvN.top200\"]], assay = \"peaks\") -> evn.peaks\n",
    "features <- list(\"EvN.top200\" = evn.peaks)\n",
    "\n",
    "# pathway scores\n",
    "def -> ghi\n",
    "\tghi <- AddChromatinModule (\n",
    "\t\tghi, features = features, genome = BSgenome.Hsapiens.UCSC.hg38\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78855340-e792-4743-ac43-4a85e80036cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "vlistViolin1 <- list()\n",
    "\n",
    "for (kp in vlist.enrich.colnames)\n",
    "{\n",
    "    plot_violin_metadata(ghi@meta.data[, c(29, 48:53)], \"peaks_snn_res.0.4\", kp) -> vlistViolin1[[kp]]\n",
    "}\n",
    "\n",
    "#violin plots for columns from metadata\n",
    "\n",
    "print_plots_in_list(vlistViolin1, title = \"VLIST1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaf5fcce-0ed1-4311-ad7e-9366dad4c4b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate heatmap of activity for groups of cells (Figure 6F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f35ed303-04a6-42a5-8ba9-bef5fd10bf46",
   "metadata": {},
   "outputs": [],
   "source": [
    "HMAP4.act.de <- function(obj, filtered_de, subset_barcodes=NULL, cell.groups=\"seurat_clusters\")\n",
    "{\n",
    "    if (!is.null(subset_barcodes)) {\n",
    "        obj <- obj[, subset_barcodes]\n",
    "    }\n",
    "    intersect(unique(filtered_de),rownames(obj@assays$RNA@data)) -> genes.heatmap\n",
    "    c(\"no.cytotox\", \"low.cytotox\", \"high.cytotox\") -> clusters.heatmap \n",
    "    heatmap.list <- list()\n",
    "    for (j in clusters.heatmap) {\n",
    "        select_cells_by <- sprintf(\"%s == %s\", cell.groups, \"j\")\n",
    "        obj@meta.data %>% filter(!! rlang::parse_expr(select_cells_by)) %>% rownames() -> cls.cells\n",
    "        obj@assays$RNA@data[genes.heatmap, cls.cells] %>% rowMeans2(useNames = FALSE) -> heatmap.list[[j]]\n",
    "    }\n",
    "    matrix(\n",
    "        unlist(heatmap.list), \n",
    "        ncol=length(clusters.heatmap), nrow=length(genes.heatmap)\n",
    "    ) %>% t() %>% scale() %>% t() -> mat.hmap\n",
    "    rownames(mat.hmap) <- genes.heatmap\n",
    "    colnames(mat.hmap) <- clusters.heatmap\n",
    "   # return(mat.hmap)\n",
    "    #mat.hmap <- mat.hmap[, c(\"0\", \"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\")]\n",
    "    col_fun_act = paletteContinuous(set = \"blueYellow\", n = 5, reverse = FALSE)\n",
    "    #col_fun_act(seq(-3, 3))\n",
    "    Heatmap(\n",
    "        mat.hmap, \n",
    "        cluster_rows=T,\n",
    "        cluster_columns=FALSE, \n",
    "        col= col_fun_act,\n",
    "        column_names_rot = 0, column_names_centered = T, row_names_gp = gpar(fontsize = 4),\n",
    "        heatmap_legend_param = list(legend_direction = \"vertical\")\n",
    "    ) -> p.n\n",
    "    #draw(p.n, heatmap_legend_side=\"bottom\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c5cfac2-d11e-4b89-9d2a-795a59940e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pdf(sprintf(\"%s/hmap.cyto.activity.pdf\", IMG_OUT), width=6, height=20)\n",
    "HMAP4.act.de(def,  filtered_de = geneset[[\"Cytotoxicity\"]], cell.groups = \"type\")\n",
    "#dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04f18c1c-2616-4112-a74e-7084ec90ece4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
